File: 0493-Reverse Pairs/0493-Reverse Pairs.cpp
class Solution {
public:
    int reversePairs(vector<int>& nums) {
        int n = nums.size();
        if (n == 0) {
            return 0;
        }
        
        int count = 0;
        vector<int> temp(n);
        mergeSort(nums, 0, n - 1, temp, count);

        return count;
    }
    
private:
    void mergeSort(vector<int>& nums, int start, int end, vector<int>& temp, int& count) {
        if (start < end) {
            int mid = start + (end - start) / 2;
            mergeSort(nums, start, mid, temp, count);
            mergeSort(nums, mid + 1, end, temp, count);
            merge(nums, start, mid, end, temp, count);
        }
    }
    
    void merge(vector<int>& nums, int start, int mid, int end, vector<int>& temp, int& count) {
        int i = start, j = mid + 1, p = mid + 1;
        int k = start;
        while (i <= mid) {
            while (p <= end && nums[i] > 2LL * nums[p]) {
                ++p;
            }
            
            count += p - (mid + 1);
            
            while (j <= end && nums[j] < nums[i]) {
                temp[k++] = nums[j++];
            }
            temp[k++] = nums[i++];
        }
        
        while (j <= end) {
            temp[k++] = nums[j++];
        }
        
        for (int i = start; i <= end; ++i) {
            nums[i] = temp[i];
        }
    }
};

File: 0401-0500/0493-Reverse Pairs/0493-Reverse Pairs.py
class Solution:
    def reversePairs(self, nums: List[int]) -> int:
        temp = [0]*len(nums)
        def mergeSort(start, end):
            count = 0
            if start < end:
                mid = (start + end) // 2
                count += mergeSort(start, mid)
                count += mergeSort(mid + 1, end)
                
                i = k = start
                j = p = mid + 1
                while i <= mid:
                    while p <= end and nums[i] > 2*nums[p]:
                        p += 1
                    count += p - (mid+1)
                    while j <= end and nums[i] > nums[j]:
                        temp[k] = nums[j]
                        k += 1
                        j += 1
                        
                    temp[k] = nums[i]
                    k += 1
                    i += 1
                    
                while j <= end:
                    temp[k] = nums[j]
                    k += 1
                    j += 1
                    
                nums[start:end+1] = temp[start:end+1]
            return count
        
        return mergeSort(0, len(nums) - 1)
